### ISSUE TITLE
Compatibility with Next.js

### ISSUE THREAD

**Issue by heracek** [supertokens/supertokens-auth-react#172]
Hi,
I love SuperTokens and I think you are doing great job!

I recently updated to new version of and it seems that [this change included in v0.5.2 introduced a breaking change](https://github.com/supertokens/supertokens-auth-react/pull/150/files#diff-a4468abdb2502816b9e2b9011227077db2ba380bff62673fbe51725b2336bad9R110-R118) for Next.js.

## My Setup

I use `yarn@1.22.10` and `next@10.0.6`. With `supertokens-auth-react@0.5.1` all works ok, with newer version it breaks.

## Problem

I'm getting this error on v0.5.2:

> Unhandled Runtime Error
> Error: Invariant failed: You should not use \<withRouter(EmailPasswordAuth) /\> outside a \<Router\>

## Analysis

- in v0.5.2 which includes #150 there is an extra check for `react-router-dom` in a wrapper for `EmailPasswordAuth` component
  - if there is `react-router-dom` installed it will automatically assume that you are using this library for routing. Which could be often wrong assumption. This dependency could be in `node_modules` for other application in case of monorepo setup.
  - if there is no `react-router-dom` it will work without it. This is what is needed for Next.js.
- secondary problem is that `react-router-dom` is placed in `optionalDependencies` of `supertokens-auth-react`
  - `yarn` will install optional dependencies by default

So situation is this:
- `yarn` install's `react-router-dom` because it is `optionalDependencies` of `supertokens-auth-react`
- `EmailPasswordAuth` will detect `react-router-dom` and wrap the component with it
- I use `EmailPasswordAuth` in Next.js
- `react-router-dom` shows error that I'm using it without proper setup

## Solution

- current solution seems with `react-router-dom` detection seems a bit hacky
- there should be more direct way to how to import `EmailPasswordAuth`:
  - one with and one without `react-router-dom`
- in a long some lower-level API could be introduced so users of SuperTokens can plug it in any SPA routing library

If you like I can help with the refactor or with finding possible solution.

**Comment by rishabhpoddar** [comment]
Thanks for pointing out this issue @heracek.

@NkxxkN I have assigned this to you. We can do the following:
- Add a config in the supertokens.init level like `useReactRouterDom` which by default is determined based on it being in node_modules. So in case the user does not want to use it, and still has it in node_modules, then they can set that to `false` and all should work fine.

@heracek will this solution solve your issue?

**Comment by heracek** [comment]
Hi,
thanks for a quick answer!

Suggested solution could help in a short term or as a quick fix.

My assumption is as SuperTokens will get more and more users it will need a support for other routers too (eg. [Reach Router](https://reach.tech/router/), [Wouter](https://github.com/molefrog/wouter/), ...).

So my suggestion is make imports of those components more explicit, e.g.:

```ts
// for usage in `next`
import { EmailPasswordAuth } from 'supertokens-auth-react/recipe/emailpassword/nextjs';
///---
// for usage with `react-router-dom`
import { EmailPasswordAuth } from 'supertokens-auth-react/recipe/emailpassword/react-router-dom';
```

Also introduction of some pluggable architecture could be a solution. For example `EmailPasswordAuth` can get configuration callbacks as a prop that will tell it what to call if it needs redirect to some URL.

Current solution is kind of "auto-magical" - it enables some ease of uses, but requires "hacks" in the background and I would say that it will not scale well.

I think configuration flags like `useReactRouterDom` would not scale well to multiple router support too.

**Comment by NkxxkN** [comment]
Hey @heracek , 
Thanks for reporting and apologise for the inconvenience. 
The use case where NextJS was used alongside `react-router-dom` wasn't one I had thought off.

We will go with the `useReactRouterDom` config for now and will think about how to handle multiple routers in a scalable way as a follow up.

This commit will solve for now https://github.com/supertokens/supertokens-auth-react/commit/95289da37349ca89737566a72495fcada0befeb8 and we will release 0.5.6 asap and let you know.

**Comment by NkxxkN** [comment]
Aditionnally, moving `react-router-dom` from `optionalDependencies` to `peerDependencies` will make sure that it's not installed when not needed. Hence that could be a fix on its own, but we will keep the `useReactRouterDom` option for extra safety.

**Comment by rishabhpoddar** [comment]
@NkxxkN I have reviewed the changes, please see my comment: https://github.com/supertokens/supertokens-auth-react/commit/95289da37349ca89737566a72495fcada0befeb8#r46823642

**Comment by heracek** [comment]
Hi @NkxxkN,
thanks for a quick fix!

I've tested running on `"supertokens-auth-react": "https://github.com/supertokens/supertokens-auth-react#0.5-supertokens-init-normalisation"` in `package.json`
and I can confirm that Next.js is working again.

Thanks!

**Comment by heracek** [comment]
I'm testing it bit more on `dev-0.5.6` and all works ok.

Only minor issue is that Next.js dev server (`yarn dev`) is now printing following warnings to terminal:
```
warn  - ./node_modules/supertokens-auth-react/lib/build/superTokens.js
Module not found: Can't resolve 'react-router-dom' in '/<path>/node_modules/supertokens-auth-react/lib/build'
```

But it's really a minor think, it can wait 😉

**Comment by rishabhpoddar** [comment]
@heracek you should use branch 0.5 (https://github.com/supertokens/supertokens-auth-react#0.5) for now. We will fix the above issue and release the version 0.5.6 soon (and notify here). Thanks

**Comment by rishabhpoddar** [comment]
@heracek we have released the version 0.5.6. Unfortunately, we cannot find a fix for the warning at the moment, but will have a look at it later.
